File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/view/TerminalTextView.java
package com.inspiredandroid.linuxcommandbibliotheca.view;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Typeface;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatTextView;
import android.text.SpannableString;
import android.text.method.LinkMovementMethod;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.util.AttributeSet;

import com.inspiredandroid.linuxcommandbibliotheca.R;
import com.inspiredandroid.linuxcommandbibliotheca.misc.FragmentCoordinator;
import com.inspiredandroid.linuxcommandbibliotheca.misc.TypefaceUtils;

/**
 * Created by Simon Schubert
 * <p/>
 * This View makes it very easy to highlightQueryInsideText defined mCommands in an normal textview. Define the
 * mCommands which should be highlighted in an string array and link it in the layout resource as
 * "command".
 */
public class TerminalTextView extends AppCompatTextView {

    private String[] mCommands;
    private int[] mOutputRows;

    public TerminalTextView(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.TerminalTextView);
        int resID = ta.getResourceId(R.styleable.TerminalTextView_commands, R.array.default_codetextview_commands);
        int outputRowsResID = ta.getResourceId(R.styleable.TerminalTextView_outputRows, R.array.default_codetextview_commands);
        mCommands = context.getResources().getStringArray(resID);
        mOutputRows = context.getResources().getIntArray(outputRowsResID);
        boolean mIgnoreTerminalStyle = ta.getBoolean(R.styleable.TerminalTextView_ignoreTerminalStyle, false);
        ta.recycle();

        if (isInEditMode()) {
            return;
        }

        updateLinks();
        setMovementMethod(LinkMovementMethod.getInstance());
        setHighlightColor(Color.TRANSPARENT);

        if (!mIgnoreTerminalStyle) {
            Typeface typeface = TypefaceUtils.getTypeFace(getContext());
            setTypeface(typeface);
        }
    }

    /**
     * Set clickable man pages(mCommands)
     *
     * @param commands
     */
    public void setCommands(String[] commands) {
        mCommands = commands;
        updateLinks();
    }

    /**
     * Mark man pages(mCommands) clickable
     */
    private void updateLinks() {
        setText(createSpannable(getText().toString(), mCommands));
    }

    /**
     * Highlights Commands of the text and make them clickable
     *
     * @param text     spannable content
     * @param commands list of mCommands to highlightQueryInsideText
     * @return
     */
    private SpannableString createSpannable(String text, String[] commands) {
        SpannableString ss = new SpannableString(text);

        for (final String command : commands) {
            ClickableTextView.addClickableSpanToPhrases(ss, text, command, () ->
                    FragmentCoordinator.startCommandManActivity(getActivity(), command));
        }


        addItalicSpans(ss, text);
        addOutputSpans(ss, text);

        return ss;
    }

    private FragmentActivity getActivity() {
        Context context = getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof FragmentActivity) {
                return (FragmentActivity) context;
            }
            context = ((ContextWrapper) context).getBaseContext();
        }
        return null;
    }

    /**
     * @param ss
     * @param text
     */
    private void addOutputSpans(SpannableString ss, String text) {
        if (mOutputRows.length == 0) {
            return;
        }

        String lines[] = text.split("\\r?\\n");
        int start = 0;
        int end = 0;

        for (int i = 0; i < lines.length; i++) {
            end += lines[i].length();
            if (doesArrayContainsInt(mOutputRows, i)) {
                if (start < 0 || end < 0 || start >= end) {
                    break;
                }
                ss.setSpan(new ForegroundColorSpan(ContextCompat.getColor(getContext(), R.color.grey)), start, end, 0);
            }
            end += 1;
            start = end;
        }
    }

    /**
     * @param array
     * @param value
     * @return
     */
    private boolean doesArrayContainsInt(int[] array, int value) {
        for (int anArray : array) {
            if (anArray == value) {
                return true;
            }
        }
        return false;
    }

    /**
     * Make placeholder text italic
     *
     * @param ss
     * @param text
     */
    private void addItalicSpans(SpannableString ss, String text) {
        int indexStart = 0;
        while (text.indexOf("[", indexStart) != -1) {
            int start = text.indexOf("[", indexStart);
            int end = text.indexOf("]", indexStart);
            if (start < 0 || end < 0 || start >= end) {
                break;
            }
            ss.setSpan(new StyleSpan(Typeface.ITALIC), start, end + 1, 0);
            indexStart = end + 1;
        }
    }
}

File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/view/TerminalTextView.kt
package com.inspiredandroid.linuxcommandbibliotheca.view

import android.content.Context
import android.content.ContextWrapper
import android.graphics.Color
import android.graphics.Typeface
import android.text.SpannableString
import android.text.method.LinkMovementMethod
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity

import com.inspiredandroid.linuxcommandbibliotheca.R
import com.inspiredandroid.linuxcommandbibliotheca.interfaces.OnClickLinkListener
import com.inspiredandroid.linuxcommandbibliotheca.misc.FragmentCoordinator

/* Copyright 2019 Simon Schubert
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
class TerminalTextView(context: Context, attrs: AttributeSet) : AppCompatTextView(context, attrs) {

    private var commands: Array<String> = arrayOf()
    private val outputRows: IntArray

    private val activity: FragmentActivity?
        get() {
            var context = context
            while (context is ContextWrapper) {
                if (context is FragmentActivity) {
                    return context
                }
                context = context.baseContext
            }
            return null
        }

    init {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.TerminalTextView)
        val resID = ta.getResourceId(R.styleable.TerminalTextView_commands, R.array.default_codetextview_commands)
        val outputRowsResID = ta.getResourceId(R.styleable.TerminalTextView_outputRows, R.array.default_codetextview_commands)
        commands = context.resources.getStringArray(resID)
        outputRows = context.resources.getIntArray(outputRowsResID)
        ta.recycle()

        if (!isInEditMode) {
            updateLinks()
            movementMethod = LinkMovementMethod.getInstance()
            highlightColor = Color.TRANSPARENT
        }
    }

    /**
     * Set clickable man pages(commands)
     *
     * @param commands
     */
    fun setCommands(commands: Array<String>) {
        this.commands = commands
        updateLinks()
    }

    /**
     * Mark man pages(commands) clickable
     */
    private fun updateLinks() {
        text = createSpannable(text.toString(), commands)
    }

    /**
     * Highlights Commands of the text and make them clickable
     *
     * @param text     spannable content
     * @param commands list of commands to highlightQueryInsideText
     * @return
     */
    private fun createSpannable(text: String, commands: Array<String>): SpannableString {
        val ss = SpannableString(text)

        for (command in commands) {
            val listener = object : OnClickLinkListener {
                override fun onClickLink() {
                    FragmentCoordinator.startCommandManActivity(activity, command)
                }
            }
            ClickableTextView.addClickableSpanToPhrases(ss, text, command, listener)
        }

        addItalicSpans(ss, text)
        addOutputSpans(ss, text)

        return ss
    }

    /**
     * @param ss
     * @param text
     */
    private fun addOutputSpans(ss: SpannableString, text: String) {
        if (outputRows.isEmpty()) {
            return
        }

        val lines = text.split("\\r?\\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var start = 0
        var end = 0

        for (i in lines.indices) {
            end += lines[i].length
            if (outputRows.any { it == i }) {
                ss.setSpan(ForegroundColorSpan(ContextCompat.getColor(context, R.color.grey)), start, end, 0)
            }
            end += 1
            start = end
        }
    }

    /**
     * Make placeholder text italic
     *
     * @param ss
     * @param text
     */
    private fun addItalicSpans(ss: SpannableString, text: String) {
        var indexStart = 0
        while (text.indexOf("[", indexStart) != -1) {
            val start = text.indexOf("[", indexStart)
            val end = text.indexOf("]", indexStart)
            if (start == -1 || end == -1 || start >= end) {
                break
            }
            ss.setSpan(StyleSpan(Typeface.ITALIC), start, end + 1, 0)
            indexStart = end + 1
        }
    }
}